hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/TinybarValues.java

```diff
@@ -115,17 +152,50 @@ public long childTransactionTinybarGasPrice() {
                         * childTransactionResourcePrices.congestionMultiplier());
     }
 
+    /**
+     * Returns the tinycent gas price for dispatching a child transaction based on the current exchange rate,
+     * Without denomination, but with congestion multiplier, saving the precision.
+     * @return the tinycent gas price
+     */
+    public long childTransactionTinycentGasPrice() {
+        if (childTransactionResourcePrices == null) {
+            throw new IllegalStateException("Cannot dispatch a child transaction from a query");
+        }
+        return childTransactionResourcePrices.basePrices().servicedataOrThrow().gas()
+                * childTransactionResourcePrices.congestionMultiplier();
+    }
+
     /**
      * Returns the tinybar-denominated price of a RAM-byte-hour (rbh) for the current operation based on the current
      * exchange rate, the current congestion multiplier, and the tinycent-denominated price of a rbh in the
      * {@code service} fee component.
      *
      * @return the tinybar-denominated price of a rbh for the current operation
      */
-    public long topLevelTinybarRbhPrice() {
-        return asTinybars(
-                topLevelResourcePrices.basePrices().servicedataOrThrow().rbh()
-                        / FEE_SCHEDULE_UNITS_PER_TINYCENT
-                        * topLevelResourcePrices.congestionMultiplier());
+    public long topLevelTinycentRbhPrice() {
```

Comment:
Name here doesn't match code at L177 or the Javadoc.

(Which kind of makes me wonder how far the mismatch extends and affects things...)

Reply:
I'll update the javadoc.
The code at l177 is the old code that is producing the precision loss because of the division there. It shall be removed with [#15098](https://github.com/hashgraph/hedera-services/issues/15098) when we verify that the calculations are working as expected.

Reply:
The conversion tinycents -> tinybars was producing the precision loss because of the divisions and the exchange rate. We actually don't need that conversions to calculate the gas but for backwards compatibility we kept the old code (that uses tinybars to calculate the gas) under the feature flag. Some of the methods are still present and only used by the `SystemContracGasCalculator` methods that are annotated as `Deprecated`.